### DISCUSSION TITLE
Regarding do_step

### DISCUSSION THREAD

**Discussion by kgvkalyan** [NTNU-IHB/PythonFMU#127] [Q&A]
I tried converting sympyslave example into fmu and tried running it in matlab simulink, it shows some error regarding the fmi2dostep error.
> I also wanted to know how we can convert a python code which has time step simulation and I want to use that fmu for co simulation. 
Thank you

**Comment by markaren** [comment]
You need to specify what the error is. Also test it in other software, like fmpy.gui. Also check that the code runs in pure Python. There might be logical errors in your code.

> I also wanted to know how we can convert a python code which has time step simulation and I want to use that fmu for co simulation.

I don't understand the question.

**Reply by kgvkalyan** [reply]
I wanted convert my script to FMU such that I can use this for co-simulation environment. I have a function which can simulate results for given step size and time. So I wanted to know how I can give this time step and total simulation time.
from pythonfmu import Fmi2Causality, Fmi2Variability, Fmi2Slave, Real
import jos3
import pandas as pd


class HBM(Fmi2Slave):

    author = "kgvksk"
    description = "Human Body Modelling"

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.To = 28 # Operating Temperature
        self.RH = 50 # Relative_humidity
        self.Va = 0.1 # Air Velocity
        self.Tskmean = 28 # Skin Temperature
        self.register_variable(Real("To", causality=Fmi2Causality.input))
        self.register_variable(Real("RH", causality=Fmi2Causality.input))
        self.register_variable(Real("Va", causality=Fmi2Causality.input))
        self.register_variable(Real("Tskmean", causality=Fmi2Causality.output))
        
    def do_step(self, current_time, step_size):
        model = H_B_M.HBM(height = 1.71, weight = 74, age = 23)
        model.simulate(60)
        df = pd.DataFrame(modeljos3.dict_results() )
        Tskmean = df["TskMean"]
        Tskmean = Tskmean.values.tolist()
        Tskmean = Tskmean[0]
        self.Tskmean = Tskmean        
        return True

So in the above code I did simulation for 60 seconds, but what I want is I want it to be simulated for every second and send data to other FMU to which it is connected.

Thankyou for you reply.

**Reply by markaren** [reply]
You implement do_step (required) and setup, enter and exit initialization mode and terminate (optional) however you see fit.

**Reply by markaren** [reply]
In this example you are being told to step the model `step_size` seconds forward in time, but you construct the model and and simulate it for 60 seconds, which is wrong.

**Reply by kgvkalyan** [reply]
I have edited the comment , I have attaced the code which I would like to convert into FMU. any suggestions?

**Reply by kgvkalyan** [reply]
I did not understand the do_step function, where does it get the time step and total time? I am really new to this FMU stuff please bear me if ask some basic questions

**Reply by markaren** [reply]
The FMI master tells the FMI slaves to step x seconds forwards in time.

**Reply by kgvkalyan** [reply]
So, I have to give additional input signals for time and time step in the code and then specify these in the FMI master.

**Reply by markaren** [reply]
It depends on the FMI master how you set the time-step. Most use a fixed-step timestep which you specify once. (If you don't specify it, it probably uses some default value). current_time is tracked internally by the master, which forwards it to your slave.

**Reply by kgvkalyan** [reply]
Thankyou so much. I'll get back to you if I have any problem.

**Reply by kgvkalyan** [reply]
![image](https://user-images.githubusercontent.com/57713233/117763993-631af580-b249-11eb-99a6-e8b32be08094.png)
I tried running the FMU in MATLAB simulink I got the above error.

**Reply by markaren** [reply]
Not much I can help with based on that error message, Try different tools etc.

**Reply by kgvkalyan** [reply]
Okay, Thankyou.
